### DISCUSSION TITLE
Enhancement: [consistent-type-assertion] Bring in options from total-functions/no-unsafe-type-assertion

### DISCUSSION THREAD

**Discussion by JoshuaKGoldberg** [typescript-eslint/typescript-eslint#6011] [RFCs]
### Before You File a Proposal Please Confirm You Have Done The Following...

- [X] I have [searched for related issues](https://github.com/typescript-eslint/typescript-eslint/issues?q=is%3Aissue+label%3A%22enhancement%3A+plugin+rule+option%22) and found none that match my proposal.
- [X] I have searched the [current rule list](https://typescript-eslint.io/rules/#supported-rules) and found no rules that match my proposal.
- [X] I have [read the FAQ](https://typescript-eslint.io/docs/linting/troubleshooting) and my problem is not listed.

### My proposal is suitable for this project

- [X] I believe my proposal would be useful to the broader TypeScript community (meaning it is not a niche proposal).

### Link to the rule's documentation

https://typescript-eslint.io/rules/consistent-type-assertions

### Description

https://github.com/danielnixon/eslint-plugin-total-functions#total-functionsno-unsafe-type-assertion has some useful options:

> * this rule is weaker than `consistent-type-assertions` with its `assertionStyle` option set to `never` -- this rule will permit type assertions that it considers safe as opposed to blanket banning all type assertions, and
> * this rule is stronger than `consistent-type-assertions` with its `objectLiteralTypeAssertions` option set to `never`

Can we add these options to `@typescript-eslint/consistent-type-assertions`?

### Fail

```typescript
type Foo = { readonly bar: number };
const foo = {};
const foo2 = foo as Foo;
foo2.bar.toString();
```

### Pass

```typescript
type Foo = { readonly bar: number };
const foo = {};
const foo2 = foo as Partial<Foo>;
foo2.bar?.toString();
```

### Additional Info

Forking conversation out from https://github.com/danielnixon/eslint-plugin-total-functions/issues/665.

**Comment by bradzacher** [comment]
TBH I really hate `consistent-type-assertions`.
A while back we merged a few rules together and I think that the UX suffered because of it.

One big problem is that you can do `assertionStyle: 'never'` - which causes the rule to ban all assertions. But then you've got no mechanism for enforcing that you use an `as` over an angle bracket assertion - which is obv a problem. A lot of users complained.

I think we need to take a look at this rule and really evaluate the point of it and what we might do better.
